### PR TITLE
Update title only when necessary

### LINKED ISSUE #122: Title should be set and only updated when Nvim sends set_title message
Currently the title is updated every time a redraw occurs. Redraws happen often and as such the title gets set when it shouldn't have to be set.  

Within the `redraw` rpc msg there is an action called `set_title` that specifies the title and is received when the title should be updated. The title should only be updated when this message is received. Likewise, the title should be set to the contents of this message. 

This `set_title` can be seen by running the Neovim binary from terminal (not using open).

### PR DESCRIPTION
**Pull request by philplckthun**
Fixes #122

### CODE REVIEW

**Comment by bambu** [comment]
I would love to hear @rogual's feedback on this subject as he is the project manager. 

Personally, I would prefer to handle all of the window logic within the window class. Something like calling `[mWindow updateWindowTitle]` with a title and having the logic that you have in the set_title case there. 

Also, should we set the window title to our own custom `titlestring` or display the one chosen by the user?

**Comment by philplckthun** [comment]
@bambu Totally didn't watch out there thanks.

**Comment by rogual** [comment]
Shouldn't the representedFileName always be the pathname of the buffer?

**Comment by rogual** [comment]
Also, one tricky detail: The standard for Mac apps is for the window to show the file icon & filename, and for the filename to be displayed with respect to the user's preferences regarding file extension visibility, i.e. `myfile.txt` or `myfile` depending on settings.

`setTitleWithRepresentedFilename` handles this logic automatically, but if we're allowing custom titles to come in from Vim, this becomes prohibitively difficult.

My suggestion: When `title` is off, use current method & set title according to HIG. When `title` is on, use your method, but still set the “represented file” (the icon) from the buffer's pathname.

Of course, this does depend on us being able to detect whether `title` is on or off. What does a blank `set_title` event signify?

**Comment by philplckthun** [comment]
I guess we should watch out for the titlestring variable, not the title variable set to on?

**Comment by bambu** [comment]
hmm.. had not explored `set notitle`. Not having a title, will make it not send the `set_title` action. In that case, maybe something like @philplckthun's original commit (`34780f6`) would work better. Sorry @philplckthun for asking that you change it the first time around.  Doing it this way will bring down the amount of times that it updates the titlebar significantly. 

The only thing that I would change from that commit is to have that `autocmd` call a vim function instead of the rpc call directly. It allows for adding more logic in the future to the bufenter autocmd (technically the 1 line can be changed in the future if the need arises).

**Comment by philplckthun** [comment]
@bambu @rogual Done. Also thought about implementing the set_title event to be processed to set the title separately, but that's really a mess with the two separate classes, synchronicity and such, so let's keep it simple for now?

**Comment by rogual** [comment]
Could you summarize what this PR is supposed to do?

**Comment by rogual** [comment]
I mean, looking at the code, it looks like this only updates the title on BufEnter, which would be a significant loss of functionality if BufEnter fires when I think it does (only when you load a new buffer)?

**Comment by philplckthun** [comment]
@rogual It fires each time you enter a buffer, not load/open/whatever a new buffer ;)

**Comment by rogual** [comment]
Aha, thanks.

Right now I'm thinking we probably souldn't merge this into master until it can handle setting the title when the buffer you're in changes its title — it seems to me like there are still a lot of situations where that could happen and it wouldn't be a good user experience to have the title remain out of date in those cases.

Also, is the intent of this to improve performance?

**Comment by philplckthun** [comment]
First of all it should improve the performance a little.

If you look at the state of this PR before the revert you'll find a working solution of implementing the set_title event. It was reverted to accommodate for the user changing the title variable. I can still reset it to that if you prefer it.

**Comment by bambu** [comment]
This should do the trick:
```vim
autocmd BufEnter,BufFilePost * call MacBufEnter()
```
Per the (events documentation)[http://vimhelp.appspot.com/autocmd.txt.html#BufFilePost] BufFilePost gets triggered after a file name is changed. Tested the above with :sav, :w and doing a SaveAs.

**Comment by philplckthun** [comment]
@bambu Pushed a new commit to use BufFilePost as well

**Comment by rogual** [comment]
I'm happy to merge this if it results in the title being up-to-date at least as reliably as it is on master.

**Comment by bambu** [comment]
How about this one.
```vim
autocmd BufEnter,BufFilePost,BufWritePost * call MacBufEnter()
```
I could not make it fail with adding `BufWritePost` to it.

**Comment by rogual** [comment]
Thanks!

**Comment by rogual** [comment]
Merged this because it appears to fix the rather nasty https://github.com/rogual/neovim-dot-app/issues/112

I'm still getting some cases where the title doesn't update but I'll open a new issue.
